File: .claude/skills/url-to-page/resolve_page.mjs
#!/usr/bin/env node
/**
 * Resolve a page URL to its Vue page component + layout + child tree.
 *
 * For Laravel apps built on this Inertia + Fortify + Wayfinder template (resolution is by
 * `artisan`, not a hardcoded project name). Node built-ins only — no npm install needed.
 */

import { execFileSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';

const PAGES_DIR = 'resources/js/pages';
const APP_TS = 'resources/js/app.ts';
const FORTIFY_PROVIDER = 'app/Providers/FortifyServiceProvider.php';

// Fortify GET "view" routes -> the FortifyServiceProvider method that registers the render.
const FORTIFY_ROUTE_TO_VIEW = {
    login: 'loginView',
    register: 'registerView',
    'password.request': 'requestPasswordResetLinkView',
    'password.reset': 'resetPasswordView',
    'verification.notice': 'verifyEmailView',
    'password.confirm': 'confirmPasswordView',
    'two-factor.login': 'twoFactorChallengeView',
};

const reEscape = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const isFile = (p) => {
    try {
        return fs.statSync(p).isFile();
    } catch {
        return false;
    }
};
const isDir = (p) => {
    try {
        return fs.statSync(p).isDirectory();
    } catch {
        return false;
    }
};
const toPosix = (p) => p.split(path.sep).join('/');

// ---------------------------------------------------------------------------
// Repo setup
// ---------------------------------------------------------------------------

function findRepoRoot(start) {
    let cur = path.resolve(start);

    for (;;) {
        if (isFile(path.join(cur, 'artisan'))) {
            return cur;
        }

        const parent = path.dirname(cur);

        if (parent === cur) {
            return null;
        }

        cur = parent;
    }
}

function read(repoRoot, rel) {
    const p = path.join(repoRoot, rel);

    return isFile(p) ? fs.readFileSync(p, 'utf8') : null;
}

// ---------------------------------------------------------------------------
// Routes
// ---------------------------------------------------------------------------

function loadRoutes(repoRoot) {
    let stdout;

    try {
        stdout = execFileSync('php', ['artisan', 'route:list', '--json'], {
            cwd: repoRoot,
            encoding: 'utf8',
            maxBuffer: 64 * 1024 * 1024,
        });
    } catch (e) {
        const msg = e.stderr || e.message || '';
        console.error(
            `ERROR: \`php artisan route:list --json\` failed:\n${msg}`,
        );
        process.exit(1);
    }

    if (!stdout.trim()) {
        console.error(
            'ERROR: `php artisan route:list --json` produced no output.',
        );
        process.exit(1);
    }

    return JSON.parse(stdout);
}

function uriToRegex(uri) {
    const parts = uri.split('/').map((seg) => {
        const m = /^\{(\w+)(\?)?\}$/.exec(seg);

        if (m) {
            return m[2] ? '[^/]*' : '[^/]+';
        }

        return reEscape(seg);
    });

    return '^/?' + parts.join('/') + '/?$';
}

function matchRoute(routes, urlPath) {
    const candidates = [];

    for (const r of routes) {
        if (!String(r.method || '').includes('GET')) {
            continue;
        }

        const uri = r.uri;

        if (new RegExp(uriToRegex(uri)).test(urlPath)) {
            candidates.push({ hasParam: uri.includes('{'), r });
        }
    }

    if (!candidates.length) {
        return null;
    }

    // static routes (no {param}) win over parameterised ones
    candidates.sort((a, b) => (a.hasParam ? 1 : 0) - (b.hasParam ? 1 : 0));

    return candidates[0].r;
}

// ---------------------------------------------------------------------------
// Component-name resolution (3 sources)
// ---------------------------------------------------------------------------

function componentFromRouteInertia(repoRoot, uri) {
    const routesDir = path.join(repoRoot, 'routes');
    const pat = /Route::inertia\(\s*['"]([^'"]+)['"]\s*,\s*['"]([^'"]+)['"]/g;
    const target = '/' + uri.replace(/^\/+/, '');
    let files;

    try {
        files = fs.readdirSync(routesDir);
    } catch {
        return null;
    }

    for (const f of files) {
        if (!f.endsWith('.php')) {
            continue;
        }

        const txt = fs.readFileSync(path.join(routesDir, f), 'utf8');

        for (const m of txt.matchAll(pat)) {
            if ('/' + m[1].replace(/^\/+/, '') === target) {
                return m[2];
            }
        }
    }

    return null;
}

function componentFromController(repoRoot, action) {
    const m = /^(App\\Http\\Controllers\\[\w\\]+)@(\w+)/.exec(action);

    if (!m) {
        return null;
    }

    const cls = m[1];
    const method = m[2];
    const rel = cls.replace(/^App\\/, 'app\\').replace(/\\/g, '/') + '.php';
    const txt = read(repoRoot, rel);

    if (!txt) {
        return null;
    }

    const fn = new RegExp('function\\s+' + reEscape(method) + '\\s*\\(').exec(
        txt,
    );

    if (!fn) {
        return null;
    }

    const rest = txt.slice(fn.index + fn[0].length);
    const rm = /Inertia::render\(\s*['"]([^'"]+)['"]/.exec(rest);

    return rm ? rm[1] : null;
}

function componentFromFortify(repoRoot, routeName) {
    const view = FORTIFY_ROUTE_TO_VIEW[routeName];

    if (!view) {
        return null;
    }

    const txt = read(repoRoot, FORTIFY_PROVIDER);

    if (!txt) {
        return null;
    }

    const re = new RegExp(
        'Fortify::' +
            reEscape(view) +
            '\\(.*?Inertia::render\\(\\s*[\'"]([^\'"]+)[\'"]',
        's',
    );
    const m = re.exec(txt);

    return m ? m[1] : null;
}

function resolveComponent(repoRoot, route) {
    const action = route.action || '';
    const name = route.name || '';

    if (action === 'Inertia\\Controller') {
        return [
            componentFromRouteInertia(repoRoot, route.uri),
            'Route::inertia()',
        ];
    }

    if (action.startsWith('App\\Http\\Controllers')) {
        return [componentFromController(repoRoot, action), action];
    }

    if (action.startsWith('Laravel\\Fortify')) {
        return [componentFromFortify(repoRoot, name), `Fortify view (${name})`];
    }

    return [null, action];
}

// ---------------------------------------------------------------------------
// Layout chain (read live from app.ts)
// ---------------------------------------------------------------------------

function layoutImports(appTs) {
    const out = {};

    for (const m of appTs.matchAll(
        /import\s+(\w+)\s+from\s+['"]([^'"]+)['"]/g,
    )) {
        out[m[1]] = m[2];
    }

    return out;
}

function resolveLayout(repoRoot, component) {
    const appTs = read(repoRoot, APP_TS);

    if (!appTs || component == null) {
        return [];
    }

    const bodyM = /layout:\s*\([^)]*\)\s*=>\s*\{(.*?)\n\s*\},/s.exec(appTs);

    if (!bodyM) {
        return [];
    }

    const body = bodyM[1];
    const imports = layoutImports(appTs);

    const cases = [...body.matchAll(/case\s+(.+?):\s*return\s+(.+?);/gs)];
    const defaultM = /default:\s*return\s+(.+?);/s.exec(body);

    const idents = (expr) => {
        expr = expr.trim();

        if (expr === 'null') {
            return [];
        }

        const arr = /^\[(.*)\]/.exec(expr);

        return arr ? arr[1].match(/\w+/g) || [] : [expr];
    };

    let chosen = null;

    for (const c of cases) {
        const pred = c[1].trim();
        const ret = c[2];
        const eq = /name\s*===\s*['"]([^'"]+)['"]/.exec(pred);
        const sw = /name\.startsWith\(\s*['"]([^'"]+)['"]/.exec(pred);

        if (eq && component === eq[1]) {
            chosen = ret;
            break;
        }

        if (sw && component.startsWith(sw[1])) {
            chosen = ret;
            break;
        }
    }

    if (chosen == null && defaultM) {
        chosen = defaultM[1];
    }

    if (chosen == null) {
        return [];
    }

    const result = [];

    for (const ident of idents(chosen)) {
        const spec = imports[ident] ?? '?';
        const p = spec.startsWith('@/')
            ? spec.replace('@/', 'resources/js/')
            : spec;
        result.push([ident, p]);
    }

    return result;
}

// ---------------------------------------------------------------------------
// Child .vue tree
// ---------------------------------------------------------------------------

// `import <clause> from '<spec>'` — clause captures default + named bindings, spec the path.
const IMPORT_RE = /import\s+(.+?)\s+from\s+['"]([^'"]+)['"]/gs;

function resolveImport(spec, importingFile, repoRoot) {
    if (spec.startsWith('@/')) {
        return path.join(repoRoot, 'resources/js', spec.slice(2));
    }

    if (spec.startsWith('.')) {
        return path.normalize(path.join(path.dirname(importingFile), spec));
    }

    return null;
}

function importedNames(clause) {
    const brace = /\{([^}]*)\}/.exec(clause);

    if (brace) {
        const names = [];

        for (let part of brace[1].split(',')) {
            part = part.trim();

            if (part) {
                names.push(part.split(' as ')[0].trim());
            }
        }

        return names;
    }

    const bare = clause.trim();

    return /^\w+$/.test(bare) ? [bare] : [];
}

function resolveBarrel(dirAbs, names) {
    const index = ['index.ts', 'index.js']
        .map((c) => path.join(dirAbs, c))
        .find((p) => isFile(p));

    if (!index) {
        return [];
    }

    const txt = fs.readFileSync(index, 'utf8');
    const exports = {};

    for (const m of txt.matchAll(
        /export\s+\{([^}]*)\}\s+from\s+['"]([^'"]+)['"]/g,
    )) {
        const spec = m[2];

        if (!spec.endsWith('.vue')) {
            continue;
        }

        for (let part of m[1].split(',')) {
            part = part.trim();

            if (!part) {
                continue;
            }

            const name = part.includes(' as ')
                ? part.split(' as ')[1].trim()
                : part;
            exports[name] = path.normalize(path.join(dirAbs, spec));
        }
    }

    return names.filter((n) => n in exports).map((n) => exports[n]);
}

function childVue(f, repoRoot, includeUi = true) {
    if (!isFile(f)) {
        return [];
    }

    const txt = fs.readFileSync(f, 'utf8');
    let out = [];

    for (const m of txt.matchAll(IMPORT_RE)) {
        const clause = m[1];
        const spec = m[2];

        if (spec.endsWith('.vue')) {
            const r = resolveImport(spec, f, repoRoot);

            if (r) {
                out.push(r);
            }

            continue;
        }

        if (spec.startsWith('@/') || spec.startsWith('.')) {
            const target = resolveImport(spec, f, repoRoot);

            if (target && isDir(target)) {
                out.push(...resolveBarrel(target, importedNames(clause)));
            }
        }
    }

    if (!includeUi) {
        out = out.filter((c) => !toPosix(c).includes('/components/ui/'));
    }

    return out;
}

function printTree(f, repoRoot, depth, maxd, seen, includeUi = true) {
    const root = path.join(repoRoot, 'resources/js');

    for (const c of childVue(f, repoRoot, includeUi)) {
        const tag = isFile(c) ? '' : '  [MISSING]';
        const rel = path.relative(root, c);
        console.log(`${'  '.repeat(depth)}- ${rel}${tag}`);

        if (depth < maxd && !seen.has(c)) {
            seen.add(c);
            printTree(c, repoRoot, depth + 1, maxd, seen, includeUi);
        }
    }
}

// ---------------------------------------------------------------------------
// URL -> path
// ---------------------------------------------------------------------------

function urlToPath(u) {
    if (u.startsWith('/')) {
        return u.split('?')[0].split('#')[0] || '/';
    }

    let s = u;

    if (!/^[a-z][a-z0-9+.-]*:\/\//i.test(s)) {
        s = 'https://' + s.replace(/^\/+/, '');
    }

    try {
        return new URL(s).pathname || '/';
    } catch {
        return '/';
    }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

function main(argv) {
    if (!argv.length) {
        console.error(
            'usage: resolve_page.mjs <url-or-path> [--depth N] [--ui] [--layout] [--all]',
        );

        return 2;
    }

    const url = argv[0];
    let depth = 4;
    const di = argv.indexOf('--depth');

    if (di !== -1) {
        depth = parseInt(argv[di + 1], 10);
    }

    const includeUi = argv.includes('--ui') || argv.includes('--all');
    const includeLayout = argv.includes('--layout') || argv.includes('--all');

    const repoRoot = findRepoRoot(process.cwd());

    if (!repoRoot) {
        console.error(
            'ERROR: not inside a Laravel repo (no `artisan` found above cwd).',
        );

        return 1;
    }

    const urlPath = urlToPath(url);

    const routes = loadRoutes(repoRoot);
    const route = matchRoute(routes, urlPath);

    console.log(`URL path : ${urlPath}`);

    if (!route) {
        console.log(
            '\nNo route matched. Check `php artisan route:list` for the URL.',
        );

        return 0;
    }

    console.log(
        `Route    : ${String(route.method).split('|')[0]} ${route.uri}` +
            (route.name ? `  (name: ${route.name})` : ''),
    );

    const [component, source] = resolveComponent(repoRoot, route);

    if (!component) {
        console.log(`Action   : ${route.action}`);
        console.log(
            "\nCould not resolve the Inertia component name from this route's action.",
        );
        console.log(
            "Open the action above and look for Inertia::render('...').",
        );

        return 0;
    }

    const pageRel = `${PAGES_DIR}/${component}.vue`;
    const pageAbs = path.join(repoRoot, pageRel);
    console.log(`Render   : '${component}'  (via ${source})`);
    console.log(`Page     : ${pageRel}${isFile(pageAbs) ? '' : '  [MISSING]'}`);

    const layout = resolveLayout(repoRoot, component);

    if (layout.length) {
        const chain = layout
            .map(([ident, p]) => `${ident} (${p})`)
            .join(' -> ');
        console.log(`Layout   : ${chain}  (auto-applied via app.ts)`);
    } else {
        console.log('Layout   : none');
    }

    if (isFile(pageAbs)) {
        console.log(`Children (depth ${depth}):`);
        printTree(pageAbs, repoRoot, 1, depth, new Set([pageAbs]), includeUi);
    }

    if (includeLayout) {
        for (const [ident, p] of layout) {
            const layAbs = path.join(repoRoot, p);

            if (isFile(layAbs)) {
                console.log(`Layout children — ${ident} (depth ${depth}):`);
                printTree(
                    layAbs,
                    repoRoot,
                    1,
                    depth,
                    new Set([layAbs]),
                    includeUi,
                );
            }
        }
    }

    return 0;
}

process.exit(main(process.argv.slice(2)));

File: .claude/skills/url-to-page/resolve_page.py
#!/usr/bin/env python3
"""Resolve a page URL to its Vue page component + layout + child tree.

For Laravel apps built on this Inertia + Fortify + Wayfinder template (regardless of the
app's own name — resolution is by `artisan`, not a hardcoded project name).

Usage:
    resolve_page.py <url-or-path> [--depth N]

This is an Inertia + Fortify + Wayfinder app (NOT Vue Router). A URL is served by a
server-side route that renders an Inertia page name; that name maps to a .vue file under
resources/js/pages. Layouts are applied centrally in resources/js/app.ts by page-name
prefix, not imported in the page itself.

The resolver is deterministic:
  1. `php artisan route:list --json` gives uri -> name + action for every route.
  2. The matched GET route's Inertia component name is found from one of three sources,
     depending on the action:
       - action == 'Inertia\\Controller'      -> Route::inertia('<uri>', '<Component>') in routes/*.php
       - action == 'App\\Http\\Controllers...' -> Inertia::render('<Component>') in that controller method
       - action is a Fortify controller        -> Fortify::<view>(... Inertia::render('<Component>')) in FortifyServiceProvider
  3. The layout chain is read live from the `layout:` resolver in resources/js/app.ts.
  4. The .vue child tree is walked, resolving '@/...' (-> resources/js) and relative imports.

No model reasoning required.
"""

import json
import os
import re
import subprocess
import sys
from urllib.parse import urlsplit

PAGES_DIR = "resources/js/pages"
APP_TS = "resources/js/app.ts"
FORTIFY_PROVIDER = "app/Providers/FortifyServiceProvider.php"

# Fortify GET "view" routes -> the FortifyServiceProvider method that registers the render.
# Stable Fortify API: the route name is fixed even if the rendered component is renamed.
FORTIFY_ROUTE_TO_VIEW = {
    "login": "loginView",
    "register": "registerView",
    "password.request": "requestPasswordResetLinkView",
    "password.reset": "resetPasswordView",
    "verification.notice": "verifyEmailView",
    "password.confirm": "confirmPasswordView",
    "two-factor.login": "twoFactorChallengeView",
}


# ---------------------------------------------------------------------------
# Repo setup
# ---------------------------------------------------------------------------


def find_repo_root(start):
    """Walk up from `start` until a dir containing `artisan` is found (Laravel root)."""
    cur = os.path.abspath(start)
    while True:
        if os.path.isfile(os.path.join(cur, "artisan")):
            return cur
        parent = os.path.dirname(cur)
        if parent == cur:
            return None
        cur = parent


def read(repo_root, rel):
    p = os.path.join(repo_root, rel)
    if not os.path.isfile(p):
        return None
    return open(p, encoding="utf-8").read()


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------


def load_routes(repo_root):
    """Return the parsed `php artisan route:list --json` array."""
    out = subprocess.run(
        ["php", "artisan", "route:list", "--json"],
        cwd=repo_root, capture_output=True, text=True,
    )
    if out.returncode != 0 or not out.stdout.strip():
        raise SystemExit(f"ERROR: `php artisan route:list --json` failed:\n{out.stderr}")
    return json.loads(out.stdout)


def uri_to_regex(uri):
    """Convert a Laravel route uri ('settings/profile', 'users/{id}', '{post?}') to a regex."""
    parts = []
    for seg in uri.split("/"):
        m = re.fullmatch(r"\{(\w+)(\?)?\}", seg)
        if m:
            parts.append(r"[^/]+" if not m.group(2) else r"[^/]*")
        else:
            parts.append(re.escape(seg))
    return "^/?" + "/".join(parts) + "/?$"


def match_route(routes, path):
    """Find the GET route whose uri matches `path`. Prefer exact (non-param) matches."""
    candidates = []
    for r in routes:
        if "GET" not in r.get("method", ""):
            continue
        uri = r["uri"]
        if re.match(uri_to_regex(uri), path):
            has_param = "{" in uri
            candidates.append((has_param, r))
    if not candidates:
        return None
    # static routes (no {param}) win over parameterised ones
    candidates.sort(key=lambda c: c[0])
    return candidates[0][1]


# ---------------------------------------------------------------------------
# Component-name resolution (3 sources)
# ---------------------------------------------------------------------------


def component_from_route_inertia(repo_root, uri):
    """Source 1: Route::inertia('<uri>', '<Component>') across routes/*.php."""
    routes_dir = os.path.join(repo_root, "routes")
    pat = re.compile(r"""Route::inertia\(\s*['"]([^'"]+)['"]\s*,\s*['"]([^'"]+)['"]""")
    target = "/" + uri.lstrip("/")
    for f in os.listdir(routes_dir):
        if not f.endswith(".php"):
            continue
        txt = open(os.path.join(routes_dir, f), encoding="utf-8").read()
        for ruri, comp in pat.findall(txt):
            if "/" + ruri.lstrip("/") == target:
                return comp
    return None


def component_from_controller(repo_root, action):
    """Source 2: App controller -> first Inertia::render('<Component>') in the method body."""
    m = re.match(r"(App\\Http\\Controllers\\[\w\\]+)@(\w+)", action)
    if not m:
        return None
    cls, method = m.group(1), m.group(2)
    rel = cls.replace("App\\", "app\\").replace("\\", "/") + ".php"
    txt = read(repo_root, rel)
    if not txt:
        return None
    fn = re.search(r"function\s+" + re.escape(method) + r"\s*\(", txt)
    if not fn:
        return None
    rest = txt[fn.end():]
    rm = re.search(r"""Inertia::render\(\s*['"]([^'"]+)['"]""", rest)
    return rm.group(1) if rm else None


def component_from_fortify(repo_root, route_name):
    """Source 3: Fortify view route -> Inertia::render in FortifyServiceProvider."""
    view = FORTIFY_ROUTE_TO_VIEW.get(route_name)
    if not view:
        return None
    txt = read(repo_root, FORTIFY_PROVIDER)
    if not txt:
        return None
    m = re.search(
        r"Fortify::" + re.escape(view) + r"\(.*?Inertia::render\(\s*['\"]([^'\"]+)['\"]",
        txt, re.DOTALL,
    )
    return m.group(1) if m else None


def resolve_component(repo_root, route):
    action = route.get("action", "")
    name = route.get("name") or ""
    if action == "Inertia\\Controller":
        return component_from_route_inertia(repo_root, route["uri"]), "Route::inertia()"
    if action.startswith("App\\Http\\Controllers"):
        return component_from_controller(repo_root, action), action
    if action.startswith("Laravel\\Fortify"):
        return component_from_fortify(repo_root, name), f"Fortify view ({name})"
    return None, action


# ---------------------------------------------------------------------------
# Layout chain (read live from app.ts)
# ---------------------------------------------------------------------------


def layout_imports(app_ts):
    """Map layout identifier -> '@/...' import path from app.ts import lines."""
    out = {}
    for ident, spec in re.findall(
        r"""import\s+(\w+)\s+from\s+['"]([^'"]+)['"]""", app_ts):
        out[ident] = spec
    return out


def resolve_layout(repo_root, component):
    """Apply the `layout:` resolver in app.ts to `component`; return [(ident, path), ...]."""
    app_ts = read(repo_root, APP_TS)
    if not app_ts or component is None:
        return []
    body_m = re.search(r"layout:\s*\([^)]*\)\s*=>\s*\{(.*?)\n\s*\},", app_ts, re.DOTALL)
    if not body_m:
        return []
    body = body_m.group(1)
    imports = layout_imports(app_ts)

    # collect (predicate, return-expr) in source order, plus default
    cases = re.findall(r"case\s+(.+?):\s*return\s+(.+?);", body, re.DOTALL)
    default_m = re.search(r"default:\s*return\s+(.+?);", body)

    def idents(expr):
        expr = expr.strip()
        if expr == "null":
            return []
        arr = re.match(r"\[(.*)\]", expr)
        names = re.findall(r"\w+", arr.group(1)) if arr else [expr]
        return names

    chosen = None
    for pred, ret in cases:
        pred = pred.strip()
        eq = re.search(r"""name\s*===\s*['"]([^'"]+)['"]""", pred)
        sw = re.search(r"""name\.startsWith\(\s*['"]([^'"]+)['"]""", pred)
        if eq and component == eq.group(1):
            chosen = ret
            break
        if sw and component.startswith(sw.group(1)):
            chosen = ret
            break
    if chosen is None and default_m:
        chosen = default_m.group(1)
    if chosen is None:
        return []

    result = []
    for ident in idents(chosen):
        spec = imports.get(ident, "?")
        path = spec.replace("@/", "resources/js/", 1) if spec.startswith("@/") else spec
        result.append((ident, path))
    return result


# ---------------------------------------------------------------------------
# Child .vue tree
# ---------------------------------------------------------------------------

VUE_IMPORT_RE = re.compile(r"""from\s+['"]([^'"]+)['"]""")


def resolve_import(spec, importing_file, repo_root):
    """Resolve an import specifier to an absolute path. Handles '@/' and relative imports.

    Reka UI wrappers are imported as folders ('@/components/ui/button') -> index.ts barrel,
    which is not a .vue file, so those are skipped by the .vue filter in child_vue().
    """
    if spec.startswith("@/"):
        return os.path.join(repo_root, "resources/js", spec[2:])
    if spec.startswith("."):
        return os.path.normpath(os.path.join(os.path.dirname(importing_file), spec))
    return None


def child_vue(f, repo_root):
    if not os.path.isfile(f):
        return []
    out = []
    for ln in open(f, encoding="utf-8"):
        m = VUE_IMPORT_RE.search(ln)
        if not m:
            continue
        spec = m.group(1)
        if not spec.endswith(".vue"):
            continue
        r = resolve_import(spec, f, repo_root)
        if r:
            out.append(r)
    return out


def print_tree(f, repo_root, depth, maxd, seen):
    root = os.path.join(repo_root, "resources/js")
    for c in child_vue(f, repo_root):
        tag = "" if os.path.isfile(c) else "  [MISSING]"
        rel = os.path.relpath(c, root)
        print(f"{'  ' * depth}- {rel}{tag}")
        if depth < maxd and c not in seen:
            seen.add(c)
            print_tree(c, repo_root, depth + 1, maxd, seen)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


def main(argv):
    if not argv:
        print("usage: resolve_page.py <url-or-path> [--depth N]", file=sys.stderr)
        return 2
    url = argv[0]
    depth = 3
    if "--depth" in argv:
        depth = int(argv[argv.index("--depth") + 1])

    repo_root = find_repo_root(os.getcwd())
    if not repo_root:
        print("ERROR: not inside a Laravel repo (no `artisan` found above cwd).", file=sys.stderr)
        return 1

    parts = urlsplit(url if "//" in url else "//" + url, scheme="https")
    path = parts.path or "/"

    routes = load_routes(repo_root)
    route = match_route(routes, path)

    print(f"URL path : {path}")
    if not route:
        print("\nNo route matched. Check `php artisan route:list` for the URL.")
        return 0

    print(f"Route    : {route['method'].split('|')[0]} {route['uri']}"
          + (f"  (name: {route['name']})" if route.get('name') else ""))

    component, source = resolve_component(repo_root, route)
    if not component:
        print(f"Action   : {route.get('action')}")
        print("\nCould not resolve the Inertia component name from this route's action.")
        print("Open the action above and look for Inertia::render('...').")
        return 0

    page_rel = f"{PAGES_DIR}/{component}.vue"
    page_abs = os.path.join(repo_root, page_rel)
    print(f"Render   : '{component}'  (via {source})")
    print(f"Page     : {page_rel}" + ("" if os.path.isfile(page_abs) else "  [MISSING]"))

    layout = resolve_layout(repo_root, component)
    if layout:
        chain = " -> ".join(f"{ident} ({path})" for ident, path in layout)
        print(f"Layout   : {chain}  (auto-applied via app.ts)")
    else:
        print("Layout   : none")

    if os.path.isfile(page_abs):
        print(f"Children (depth {depth}):")
        print_tree(page_abs, repo_root, 1, depth, {page_abs})
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
